---
title: "MPRA_DESeq_LinearModeling"
author: "Divya Kolakada"
date: "2023-08-07"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# MPRA Analysis

This document contains the code used to process and analyze MPRA count data from sequencing libraries derived from both EJC-independent and EJC-enhanced reporter cell lines. The data underwent initial processing with DESeq2 to calculate the log2 fold change of each 10mer investigated between DMSO and SMG1i (NMD inhibited) treated samples. The log2 fold changes of sequences predicted to undergo NMD (i.e. 10mers containing a premature termination codon (PTC)) were normalized against NMD-negative sequences (10mers containing a TGG instead of a stop codon). Linear modeling was then employed to explore the relationship between sequence identity and the resulting log2 fold changes for each 10mer.

Within the code, "ART" refers to the EJC-independent data and "ARTi" refers to the EJC-enhanced data.

## Preparing The Data

```{r echo=TRUE}
#Loading all relevant libraries
library(tidyverse)
library(myPackage) #converts codons to amino acids
library(ggpubr)
library(RColorBrewer)
library(reshape2)
library(ggthemes)
library(ggplot2)
library(DESeq2)
library(broom)
library(ggrepel)
library(GGally)

# Importing txt files containing RNA count data
file_list = list.files( pattern="*.txt")

samples <- list() 
column_names <- c("Reads", "Sequence")  

# Loop over each file in file_list and read it into a data frame
for (file in file_list) {
  file_path <- 
    file.path(getwd(), file)  # Get the full file path
  samples[[file]] <- 
    read.table(file_path, 
               header = FALSE, 
               col.names = column_names)  # Read the file into a data frame with custom column names
}

# Creating art and arti data frames with counts
# ART
art <- Reduce(function(x, y) full_join(x, y, by = "Sequence"), samples[1:6]) # joining dmso and smg1i counts for art samples

colnames(art)<- c("d1", "Sequence", "d2", "d3", "s1", "s2", "s3") # d1 (dmso rep 1), s1 (smg1i rep 1)

art <- art[, c(2, 1, 3:7)] # rearranging the data frame 

# ARTi
arti <- Reduce(function(x, y) full_join(x, y, by = "Sequence"), samples[7:12]) # joining dmso and smg1i counts for arti samples

colnames(arti)<- c("d1", "Sequence", "d2", "d3", "s1", "s2", "s3") # d1 (dmso rep 1), s1 (smg1i rep 1)

arti <- arti[, c(2, 1, 3:7)] # rearranging the data frame 


# Making a histogram using the sum of all reads to get an idea of sequence representation
# ART
art_hist <- ggplot(data.frame(Sum = rowSums(art[, 2:7], na.rm = T)), aes(x = log(Sum, 10))) +
  geom_histogram(binwidth = 0.1, fill = "lightgrey", color = "black") +
  labs(title = "ART Representation Histogram",
       x = "log(Reads per Context)",
       y = "No. of Unique Contexts")

# ARTi
arti_hist <- ggplot(data.frame(Sum = rowSums(arti[, 2:7], na.rm = T)), aes(x = log(Sum, 10))) +
  geom_histogram(binwidth = 0.1, fill = "lightgrey", color = "black") +
  labs(title = "ARTi Representation Histogram",
       x = "log(Reads per Context)",
       y = "No. of Unique Contexts")

# Filter out sequences with less than a sum of 50 reads across samples
# ART
art_f<- art %>%
  filter(d1 + d2 + d3 + s1 +s2+s3 >= 50)

# ARTi
arti_f<- arti %>%
  filter(d1 + d2 + d3 + s1 +s2+s3 >= 50)
```

## DESeq2 Analysis

```{r echo=TRUE}
# DESeq2
# Create meta data for the ART and ARTi data frames. The metadata just tells you what each column other than "Sequence" is. 
# ART
art_metadata <- data.frame(
  SampleID = colnames(art_f)[-1],
  Condition = rep(c("dmso", "smg1i"), each = 3),  # 3 reps
  Replicate = rep(1:3, times = 2)  # 3 reps 
)

# ARTi
arti_metadata <- data.frame(
  SampleID = colnames(arti_f)[-1],
  Condition = rep(c("dmso", "smg1i"), each = 3),  # 3 reps
  Replicate = rep(1:3, times = 2)  # 3 reps 
)

# Running DESeq2 analysis for ART and ARTi
# ART  
art_metadata$Condition <- relevel(factor(art_metadata$Condition), ref = "smg1i") # re-level art to make SMG1i samples the reference

# Generate a list of all the 10mers with TGG at the PTC position in art
tgg_art <- subset(art_f,
                    substr(Sequence, 7, 9) %in% "TGG")
tgg_list_art<- tgg_art$Sequence

# Design experiment for DESeq2
art_dds <- DESeqDataSetFromMatrix(countData=art_f, 
                              colData=art_metadata, 
                              design=~Condition, tidy = TRUE)

art_dds<- estimateSizeFactors(art_dds, 
                               controlGenes = rownames(art_dds) %in%
                                 tgg_list_art) # normalize fold changes to TGG sequences

art_dds <- DESeq(art_dds) #running DESeq2

art_results<- results(art_dds) 

# Validating the ART DESeq data 
hist(art_results$log2FoldChange, breaks = 80) # histogram to check the distribution of the data 

art_results_d<- as.data.frame(art_results) # converting DESeq results to a data frame

art_results_d$Sequence<- rownames(art_results_d) # making rownames i.e. 10mer identity a part of the data frame

rownames(art_results_d) <- NULL #removing row names

# Converting the 10mers to codons and amino acids using frame 1
art_results_d$Codon1 <- substring(art_results_d$Sequence,1,3) # codon -2 to the PTC

art_results_d$Codon2 <- substring(art_results_d$Sequence,4,6) # codon -1 to the PTC

art_results_d$StopCodon <- substring(art_results_d$Sequence,7,9) # PTC

art_results_d$LastNt <- substring(art_results_d$Sequence,10,10) # Nucleotide immediately after the PTC

art_results_d$AA1 <- codon_to_aa( art_results_d$Codon1, seq_type = "DNA") # amino acid -2 to the PTC

art_results_d$AA2 <- codon_to_aa( art_results_d$Codon2, seq_type = "DNA") # amino acid -1 to the PTC 

# Checking whether biomodal distribution can be explained by PTC identity
art_subset_TGG<- subset(art_results_d, StopCodon=="TGG") # creating a subset from the data frame with 10mers containing TGG in the position of the PTC  

art_subset_stop<- subset(art_results_d, StopCodon=="TAG"|
                            StopCodon=="TGA" |
                            StopCodon=="TAA") # creating a subset from the data frames with 10mers containing PTCs 

ggplot()+
  geom_histogram(data=art_results_d, aes(x = log2FoldChange), fill ="grey", alpha = 1, binwidth = 0.1)+ # all data
  geom_histogram(data=art_subset_TGG, aes(x = log2FoldChange), fill = "#0726B9", alpha = 0.6, binwidth = 0.1)+ # overlay TGG subset
  geom_histogram(data=art_subset_stop, aes(x = log2FoldChange), fill = "#CD3E34", alpha = 0.6, binwidth = 0.1)+ # overlay PTC subset
  theme_minimal()+
  xlab("Transcript Levels (DMSO/SMG1i)")+
  ylab("Count")+
  theme_classic(base_size=24)

# Checking whether data subsets separate based on identity of the amino acid -1 to the PTC
art_subset_Gly<- subset(art_results_d, AA2=="Gly") # subset of 10mers with Gly -1 to the PTC

art_subset_Tyr<- subset(art_results_d, AA2=="Tyr") # subset of 10mers with Tyr -1 to the PTC

art_aa2_hist <- ggplot()+
  geom_histogram(data=art_results_d, aes(x = log2FoldChange), fill ="grey", alpha = 1, binwidth = 0.1)+ # all data
  geom_histogram(data=art_subset_Gly, aes(x = log2FoldChange), fill = "blue", alpha = 0.4, binwidth = 0.1)+ # overlay Gly subset
  geom_histogram(data=art_subset_Tyr, aes(x = log2FoldChange), fill = "red", alpha = 0.4, binwidth = 0.1)+ # overlay Tyr subset
  theme_minimal()+
  ggtitle("EJC-Ind -1 AA")


# ARTi 
arti_metadata$Condition <- relevel(factor(arti_metadata$Condition), ref = "smg1i") # re-level arti to make SMG1i samples the reference

# Generate a list of all the 10mers with TGG at the PTC position in arti
tgg_arti <- subset(arti_f,
                    substr(Sequence, 7, 9) %in% "TGG")

tgg_list_arti<- tgg_arti$Sequence

# Design experiment for DESeq
arti_dds <- DESeqDataSetFromMatrix(countData=arti_f, 
                              colData=arti_metadata, 
                              design=~Condition, tidy = TRUE)

arti_dds<- estimateSizeFactors(arti_dds, 
                               controlGenes = rownames(arti_dds) %in%
                                 tgg_list_arti) # normalize fold changes to TGG sequences

arti_dds <- DESeq(arti_dds) #running DESeq2

arti_results<- results(arti_dds)

# Validating the ARTi DESeq data 
hist(arti_results$log2FoldChange, breaks = 80) # histogram to check the distribution of the data

arti_results_d<- as.data.frame(arti_results) # converting to data frame

arti_results_d$Sequence<- rownames(arti_results_d) # converting DESeq results to a data frame

rownames(arti_results_d) <- NULL #removing row names

# Converting the 10mers to codons and amino acids using frame 1
arti_results_d$Codon1 <- substring(arti_results_d$Sequence,1,3) # codon -2 to the PTC

arti_results_d$Codon2 <- substring(arti_results_d$Sequence,4,6) # codon -1 to the PTC

arti_results_d$StopCodon <- substring(arti_results_d$Sequence,7,9) # PTC

arti_results_d$LastNt <- substring(arti_results_d$Sequence,10,10) # nucleotide immediately after the PTC

arti_results_d$AA1 <- codon_to_aa( arti_results_d$Codon1, seq_type = "DNA") # amino acid -2 to the PTC

arti_results_d$AA2 <- codon_to_aa( arti_results_d$Codon2, seq_type = "DNA") # amino acid -1 to the PTC

# Checking whether biomodal distribution can be explained by identity of the codon in the PTC position
arti_subset_TGG<- subset(arti_results_d, StopCodon=="TGG") # creating a subset from the data frame with 10mers containing TGG in the position of the PTC

arti_subset_stop<- subset(arti_results_d, StopCodon=="TAG"|
                            StopCodon=="TGA" |
                            StopCodon=="TAA") # creating a subset from the data frames with 10mers containing PTCs

ggplot()+
  geom_histogram(data=arti_results_d, aes(x = log2FoldChange), fill ="darkgrey", alpha = 1, binwidth = 0.1)+ # all data
  geom_histogram(data=arti_subset_TGG, aes(x = log2FoldChange), fill = "#0726B9", alpha = 0.6, binwidth = 0.1)+ # overlay TGG subset
  geom_histogram(data=arti_subset_stop, aes(x = log2FoldChange), fill = "#CD3E34", alpha = 0.6, binwidth = 0.1)+ # overlay PTC subset
  theme_minimal()+
  xlab("Transcript Levels (DMSO/SMG1i)")+
  ylab("Count")+
  theme_classic(base_size = 24)

# Checking whether data subsets separate based on identity of the amino acid -1 to the PTC
arti_subset_Gly<- subset(arti_results_d, AA2=="Gly") # subset of 10mers with Gly -1 to the PTC

arti_subset_Tyr<- subset(arti_results_d, AA2=="Tyr") # subset of 10mers with Tyr -1 to the PTC

arti_aa2_hist <- ggplot()+
  geom_histogram(data=arti_results_d, aes(x = log2FoldChange), fill ="grey", alpha = 1, binwidth = 0.1)+ # all data
  geom_histogram(data=arti_subset_Gly, aes(x = log2FoldChange), fill = "blue", alpha = 0.4, binwidth = 0.1)+ # overlay Gly subset
  geom_histogram(data=arti_subset_Tyr, aes(x = log2FoldChange), fill = "red", alpha = 0.4, binwidth = 0.1)+ # overlay Tyr subset
  theme_minimal()+
  ggtitle("EJC-en -1 AA")

# Exporting data for easy access
write_csv(art_results_d, "DESeq_RNA_ART.csv") # write csv of data frame for ART

write_csv(arti_results_d, "DESeq_RNA_ARTi.csv") # write csv of data frame for ARTi

```

## Linear Modeling 

```{r echo=TRUE}
# Linear modeling: Models and dot charts were made with the same code applied to either the ART or the ARTi data set. 

# How does identity of the codon in the PTC position affect log2FC?  
DESeq_RNA_ART$StopCodon <- relevel(x = factor(DESeq_RNA_ART$StopCodon), ref = "TGG") # make the reference TGG, sequences that are not predicted to be affected by NMD

fit_stop<- lm(log2FoldChange ~ StopCodon, data = DESeq_RNA_ART) # create the model

tf_stop <- tidy(fit_stop)[-1,] # create a tidy object from the results for ease of viewing 

tf_stop$sig <- ifelse(tf_stop$p.value < 0.01, "sig", "n.s.") # add a significance threshold of p< 0.01

tf_stop$Stp<- substr(tf_stop$term, 10,12) # create column with desired substring (i.e. codon identity) for plotting 

ggdotchart(data = tf_stop, x = "Stp", y = "estimate", color = "sig", size=7)+
  xlab("Stop Codon")+
  ylab("Coefficient")+
  ylim(-2.5,0)+
  scale_color_manual( values = c("sig" = "#CF2F2E", "n.s." = "black"), name = "sig")+
  theme(text = element_text(size=24),
        axis.text.x = element_text(angle = 0, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" ) # dot chart to view the data 

#From this point, the sequences with TGG in the stop codon position were filtered out. 

d <- DESeq_RNA_ART %>% filter(StopCodon != "TGG") # Filter out all sequences with TGG at the position of the stop codon in the ART data set 

di<- DESeq_RNA_ARTi %>% filter(StopCodon != "TGG") # Filter out all sequences with TGG at the position of the stop codon in the ARTi data set

# How does identity of the amino acid in the position -1 to the PTC affect log2FC? 
d$AA2 <- relevel(x = factor(d$AA2), ref = "Ala") # make reference amino acid Alanine  

fit_AA2 <- lm(log2FoldChange ~ AA2, data = d) # create model 

tf_AA2 <- tidy(fit_AA2)[-1,] # create a tidy object from the results for ease of viewing 

tf_AA2$sig <- ifelse(tf_AA2$p.value < 0.01, "sig", "n.s.") # add a significance threshold of p< 0.01

tf_AA2$AA<- substr(tf_AA2$term,4,6) # create column with desired substring (i.e. amino acid identity) for plotting

ggdotchart(data = tf_AA2, x = "AA", y = "estimate", color = "sig", size = 5) + 
  xlab("-1 Amino Acid")+
  ylab("Coefficient")+
  ylim(-0.5,0.4)+
  scale_color_manual( values = c("sig" = "#CF2F2E", "n.s." = "black"), name = "sig")+
  theme(text = element_text(size=20),
        axis.text.x = element_text(angle = 90, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" ) # dot chart to view the data

# How does identity of the amino acid in the position -2 to the PTC affect log2FC? 
d$AA1 <- relevel(x = factor(d$AA1), ref = "Ala") # make reference amino acid Alanine

fit_AA1 <- lm(log2FoldChange ~ AA1, data = d) # create model 

tf_AA1 <- tidy(fit_AA1)[-1,] # create a tidy object from the results for ease of viewing

tf_AA1$sig <- ifelse(tf_AA1$p.value < 0.01, "sig", "n.s.") # add a significance threshold of p< 0.01

tf_AA1$AA<- substr(tf_AA1$term,4,6) # create column with desired substring (i.e. amino acid identity) for plotting

ggdotchart(data = tf_AA1, x = "AA", y = "estimate", color = "sig", size = 5) + 
  xlab("-2 Amino Acid")+
  ylab("Coefficient")+
  ylim(-0.5,0.4)+
  scale_color_manual( values = c("sig" = "#CF2F2E", "n.s." = "black"), name = "sig")+
  theme(text = element_text(size=20),
        axis.text.x = element_text(angle = 90, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" ) # dot chart to view the data

# How does the identity of the last nucleotide affect log2FC?  
fit_lnt <- lm(log2FoldChange ~ LastNt, data = d) # create model, default reference is A (adenine)

tf_lnt <- tidy(fit_lnt)[-1,] # create a tidy object from the results for ease of viewing

tf_lnt$sig <- ifelse(tf_lnt$p.value < 0.01, "sig", "n.s.") # add a significance threshold of p< 0.01

tf_lnt$lnt<- substr(tf_lnt$term,7,7) # create column with desired substring (i.e. last nt identity) for plotting

ggdotchart(data = tf_lnt, x = "lnt", y = "estimate", color = "sig", size = 7) + 
  xlab("LastNt")+
  ylab("Coefficient")+
  ylim(-0.5,0.4)+
  scale_color_manual( values = c("sig" = "#CF2F2E", "n.s." = "black"), name = "sig")+
  theme(text = element_text(size=24),
        axis.text.x = element_text(angle = 0, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" ) # dot chart to view the data

# How does identity of the codon in the position -1 to the PTC affect log2FC? 
fit_c2<- lm(log2FoldChange~ Codon2, data=d) # create model, default reference is AAA

tidy_c2<- tidy(fit_c2)[-1,] # create a tidy object from the results for ease of viewing

tidy_c2$C2 <- substring(tidy_c2$term,7,9) # create column with desired substring (i.e. -1 codon) for plotting

tidy_c2$AA2 <- codon_to_aa(tidy_c2$C2, seq_type = "DNA") # convert codon to amino acid 

tidy_c2$sig <- ifelse(tidy_c2$p.value < 0.01, "*", "") # add a significance threshold of p< 0.01

heatmap_plot_c2 <- ggplot(tidy_c2, aes(x = C2, y=1, fill = estimate)) +
  geom_tile() +
  labs(x = "Codon2", fill = "coefficient") +
  theme_minimal()+
  facet_wrap(~AA2, scales="free_x", nrow=5)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),  
        panel.grid = element_blank(),  
        )+
  geom_text(aes(label = format(sig, nsmall = 3)))+
  scale_fill_gradient(low = "#CF2F2E", high = "#0726B9")+
  ggtitle("-1 Codon: EJC-Indep") # heatmap to view the data

# How does identity of the codon in the position -2 to the PTC affect log2FC? 
fit_c1<- lm(log2FoldChange~ Codon1, data=d) # create model, default reference is AAA

tidy_c1<- tidy(fit_c1)[-1,] # create a tidy object from the results for ease of viewing

tidy_c1$C1 <- substring(tidy_c1$term,7,9) # create column with desired substring (i.e. -2 codon) for plotting

tidy_c1$AA1 <- codon_to_aa(tidy_c1$C1, seq_type = "DNA") # convert codon to amino acid

tidy_c1$sig <- ifelse(tidy_c1$p.value < 0.01, "*", "") # add a significance threshold of p< 0.01

heatmap_plot_c1 <- ggplot(tidy_c1, aes(x = C1, y=1, fill = estimate)) +
  geom_tile() +
  labs(x = "Codon1", fill = "coefficient") +
  theme_minimal()+
  facet_wrap(~AA1, scales="free_x", nrow=5)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),  
        panel.grid = element_blank(),  
        )+
  geom_text(aes(label = format(sig, nsmall = 3)))+
  scale_fill_gradient(low = "#CF2F2E", high = "#0726B9")+
  ggtitle("-2 Codon: EJC-Indep") # heatmap to view the data


```

