---
title: "MPRA_DESeq"
author: "Divya Kolakada"
date: "2023-08-07"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## MPRA DESeq Analysis

```{r deseq rna}
library(tidyverse)
library(myPackage)
library(ggpubr)
library(RColorBrewer)
library(reshape2)
library(ggthemes)
library(ggplot2)
library(DESeq2)
library(mixtools)
library(broom)
library(PWMEnrich)

#Plasmid Libraries 
art_plasmid<- ML1_CKDL220025032_1A_HK7KJDSX5_L2_correctOrientation 
art_pl_count<- data.frame(table(art_plasmid$X1))
art_pl_hist<- ggplot(art_pl_count, aes(x=log(Freq, 10)))+
  geom_histogram(binwidth = 0.05, fill = "lightgrey", color = "black") +
  labs(title = "ART Plasmid Representation Histogram",
       x = "log(Reads per Context)",
       y = "No. of Unique Contexts")+
  geom_vline(xintercept=1.865, linetype = "dashed", color = "red" )

arti_plasmid<- DK_ARTi_ARTi_CKDL220031313_1A_HKMHNDSX5_L1_correctOrientation 
arti_pl_count<- data.frame(table(arti_plasmid$X1))
arti_pl_hist<- ggplot(arti_pl_count, aes(x=log(Freq, 10)))+
  geom_histogram(binwidth = 0.05, fill = "lightgrey", color = "black") +
  labs(title = "ARTi Plasmid Representation Histogram",
       x = "log(Reads per Context)",
       y = "No. of Unique Contexts")+
  geom_vline(xintercept=2.012, linetype = "dashed", color = "red" )


##Importing txt RNA files. 
file_list = list.files( pattern="*.txt")

samples <- list() 
column_names <- c("Reads", "Sequence")  

# Loop over each file in file_list and read it into a data frame
for (file in file_list) {
  file_path <- 
    file.path(getwd(), file)  # Get the full file path
  samples[[file]] <- 
    read.table(file_path, 
               header = FALSE, 
               col.names = column_names)  # Read the file into a data frame with custom column names
}

##Creating art and arti tables with counts
art <- Reduce(function(x, y) full_join(x, y, by = "Sequence"), samples[1:6])
colnames(art)<- c("d1", "Sequence", "d2", "d3", "s1", "s2", "s3")
art <- art[, c(2, 1, 3:7)]

arti <- Reduce(function(x, y) full_join(x, y, by = "Sequence"), samples[7:12])
colnames(arti)<- c("d1", "Sequence", "d2", "d3", "s1", "s2", "s3")
arti <- arti[, c(2, 1, 3:7)]

##Making a histogram using the sum of all reads
art_hist <- ggplot(data.frame(Sum = rowSums(art[, 2:7], na.rm = T)), aes(x = log(Sum, 10))) +
  geom_histogram(binwidth = 0.1, fill = "lightgrey", color = "black") +
  labs(title = "ART Representation Histogram",
       x = "log(Reads per Context)",
       y = "No. of Unique Contexts")

arti_hist <- ggplot(data.frame(Sum = rowSums(arti[, 2:7], na.rm = T)), aes(x = log(Sum, 10))) +
  geom_histogram(binwidth = 0.1, fill = "lightgrey", color = "black") +
  labs(title = "ARTi Representation Histogram",
       x = "log(Reads per Context)",
       y = "No. of Unique Contexts")

##Filter out sequences with less than a sum of 50 reads across samples
art_f<- art %>%
  filter(d1 + d2 + d3 + s1 +s2+s3 >= 50)

arti_f<- arti %>%
  filter(d1 + d2 + d3 + s1 +s2+s3 >= 50)

##DESeq
#First creat meta data for the art and arti dataframes. 
#This metadata just tells you what each column other than "Sequence" is. 
art_metadata <- data.frame(
  SampleID = colnames(art_f)[-1],
  Condition = rep(c("dmso", "smg1i"), each = 3),  # 3 reps
  Replicate = rep(1:3, times = 2)  # 3 reps 
)

arti_metadata <- data.frame(
  SampleID = colnames(arti_f)[-1],
  Condition = rep(c("dmso", "smg1i"), each = 3),  # 3 reps
  Replicate = rep(1:3, times = 2)  # 3 reps 
)

#Create a deseq object and perform analysis for art and arti tables 
##ART DESeq analysis 
#re-level art
art_metadata$Condition <- relevel(factor(art_metadata$Condition), ref = "smg1i")

#generate a list of all the sequences in art
tgg_art <- subset(art_f,
                    substr(Sequence, 7, 9) %in% "TGG")
tgg_list_art<- tgg_art$Sequence

#design experiment for DESeq
art_dds <- DESeqDataSetFromMatrix(countData=art_f, 
                              colData=art_metadata, 
                              design=~Condition, tidy = TRUE)
#Set TGG
art_dds<- estimateSizeFactors(art_dds, 
                               controlGenes = rownames(art_dds) %in%
                                 tgg_list_art)
art_dds <- DESeq(art_dds)
art_results<- results(art_dds)

#histogram to make sure things make sense
hist(art_results$log2FoldChange, breaks = 80)

#overlaying histograms with specific sequences
art_results_d<- as.data.frame(art_results) #converting to dataframe
art_results_d$Sequence<- rownames(art_results_d) #making rownames part of the dataframe
rownames(art_results_d) <- NULL #making rownames null

art_results_d$Codon1 <- substring(art_results_d$Sequence,1,3)
art_results_d$Codon2 <- substring(art_results_d$Sequence,4,6)
art_results_d$StopCodon <- substring(art_results_d$Sequence,7,9)
art_results_d$LastNt <- substring(art_results_d$Sequence,10,10)

art_results_d$AA1 <- codon_to_aa( art_results_d$Codon1, seq_type = "DNA")
art_results_d$AA2 <- codon_to_aa( art_results_d$Codon2, seq_type = "DNA")

#Checking if this works for the stop codons
art_subset_TGG<- subset(art_results_d, StopCodon=="TGG")
art_subset_stop<- subset(art_results_d, StopCodon=="TAG"|
                            StopCodon=="TGA" |
                            StopCodon=="TAA")

theme_set(theme_classic(base_size = 24))

ggplot()+
  geom_histogram(data=art_results_d, aes(x = log2FoldChange), fill ="grey", alpha = 1, binwidth = 0.1)+
  geom_histogram(data=art_subset_TGG, aes(x = log2FoldChange), fill = "#0726B9", alpha = 0.6, binwidth = 0.1)+
  geom_histogram(data=art_subset_stop, aes(x = log2FoldChange), fill = "#CD3E34", alpha = 0.6, binwidth = 0.1)+
  theme_minimal()+
  xlab("Transcript Levels (DMSO/SMG1i)")+
  ylab("Count")+
  theme_classic(base_size=24)

  

#Checking if this works for Gly and Tyr
art_subset_Gly<- subset(art_results_d, AA2=="Gly")
art_subset_Tyr<- subset(art_results_d, AA2=="Tyr")

art_aa2_hist <- ggplot()+
  geom_histogram(data=art_results_d, aes(x = log2FoldChange), fill ="grey", alpha = 1, binwidth = 0.1)+
  geom_histogram(data=art_subset_Gly, aes(x = log2FoldChange), fill = "blue", alpha = 0.4, binwidth = 0.1)+
  geom_histogram(data=art_subset_Tyr, aes(x = log2FoldChange), fill = "red", alpha = 0.4, binwidth = 0.1)+
  theme_minimal()+
  ggtitle("EJC-Ind -1 AA")


##ARTi DESeq analysis
#re-level arti
arti_metadata$Condition <- relevel(factor(arti_metadata$Condition), ref = "smg1i")

#generate a list of all the sequences in arti 
tgg_arti <- subset(arti_f,
                    substr(Sequence, 7, 9) %in% "TGG")
tgg_list_arti<- tgg_arti$Sequence

#design experiment for DESeq
arti_dds <- DESeqDataSetFromMatrix(countData=arti_f, 
                              colData=arti_metadata, 
                              design=~Condition, tidy = TRUE)
#Set TGG
arti_dds<- estimateSizeFactors(arti_dds, 
                               controlGenes = rownames(arti_dds) %in%
                                 tgg_list_arti)
arti_dds <- DESeq(arti_dds)
arti_results<- results(arti_dds)

#histogram to make sure things make sense
hist(arti_results$log2FoldChange, breaks = 80)

#overlaying histograms with specific sequences
arti_results_d<- as.data.frame(arti_results) #converting to dataframe
arti_results_d$Sequence<- rownames(arti_results_d) #making rownames part of the dataframe
rownames(arti_results_d) <- NULL #making rownames null

arti_results_d$Codon1 <- substring(arti_results_d$Sequence,1,3)
arti_results_d$Codon2 <- substring(arti_results_d$Sequence,4,6)
arti_results_d$StopCodon <- substring(arti_results_d$Sequence,7,9)
arti_results_d$LastNt <- substring(arti_results_d$Sequence,10,10)

arti_results_d$AA1 <- codon_to_aa( arti_results_d$Codon1, seq_type = "DNA")
arti_results_d$AA2 <- codon_to_aa( arti_results_d$Codon2, seq_type = "DNA")

#Checking if this works for the stop codons
arti_subset_TGG<- subset(arti_results_d, StopCodon=="TGG")
arti_subset_stop<- subset(arti_results_d, StopCodon=="TAG"|
                            StopCodon=="TGA" |
                            StopCodon=="TAA")

ggplot()+
  geom_histogram(data=arti_results_d, aes(x = log2FoldChange), fill ="darkgrey", alpha = 1, binwidth = 0.1)+
  geom_histogram(data=arti_subset_TGG, aes(x = log2FoldChange), fill = "#0726B9", alpha = 0.6, binwidth = 0.1)+
  geom_histogram(data=arti_subset_stop, aes(x = log2FoldChange), fill = "#CD3E34", alpha = 0.6, binwidth = 0.1)+
  theme_minimal()+
  xlab("Transcript Levels (DMSO/SMG1i)")+
  ylab("Count")+
  theme_classic(base_size = 24)

#Checking if this works for Gly and Tyr
arti_subset_Gly<- subset(arti_results_d, AA2=="Gly")
arti_subset_Tyr<- subset(arti_results_d, AA2=="Tyr")

arti_aa2_hist <- ggplot()+
  geom_histogram(data=arti_results_d, aes(x = log2FoldChange), fill ="grey", alpha = 1, binwidth = 0.1)+
  geom_histogram(data=arti_subset_Gly, aes(x = log2FoldChange), fill = "blue", alpha = 0.4, binwidth = 0.1)+
  geom_histogram(data=arti_subset_Tyr, aes(x = log2FoldChange), fill = "red", alpha = 0.4, binwidth = 0.1)+
  theme_minimal()+
  ggtitle("EJC-en -1 AA")

write_csv(art_results_d, "DESeq_RNA_ART.csv")
write_csv(arti_results_d, "DESeq_RNA_ARTi.csv")

```

```{r deseq dna}
##Importing txt DNA files. 
file_list = list.files( pattern="*.txt")

samples <- list() 
column_names <- c("Reads", "Sequence")  

# Loop over each file in file_list and read it into a data frame
for (file in file_list) {
  file_path <- 
    file.path(getwd(), file)  # Get the full file path
  samples[[file]] <- 
    read.table(file_path, 
               header = FALSE, 
               col.names = column_names)  # Read the file into a data frame with custom column names
}

##Creating art and arti tables with counts

art <- Reduce(function(x, y) full_join(x, y, by = "Sequence"), samples[1:6])
colnames(art)<- c("d1", "Sequence", "d2", "d3", "s1", "s2", "s3")
art <- art[, c(2, 1, 3:7)]

arti <- Reduce(function(x, y) full_join(x, y, by = "Sequence"), samples[7:12])
colnames(arti)<- c("d1", "Sequence", "d2", "d3", "s1", "s2", "s3")
arti <- arti[, c(2, 1, 3:7)]

##Making a histogram using the sum of all reads
art_hist <- ggplot(data.frame(Sum = rowSums(art[, 2:7], na.rm = T)), aes(x = log(Sum, 10))) +
  geom_histogram(binwidth = 0.1, fill = "lightgrey", color = "black") +
  labs(title = "ART Representation Histogram",
       x = "log(Reads per Context)",
       y = "No. of Unique Contexts")

arti_hist <- ggplot(data.frame(Sum = rowSums(arti[, 2:7], na.rm = T)), aes(x = log(Sum, 10))) +
  geom_histogram(binwidth = 0.1, fill = "lightgrey", color = "black") +
  labs(title = "ARTi Representation Histogram",
       x = "log(Reads per Context)",
       y = "No. of Unique Contexts")

##Filter out sequences with less than a sum of 50 reads across samples
art_f<- art %>%
  filter(d1 + d2 + d3 + s1 +s2+s3 >= 50)

arti_f<- arti %>%
  filter(d1 + d2 + d3 + s1 +s2+s3 >= 50)

##DESeq DNA
#First create meta data for the art and arti dataframes. 
#This metadata just tells you what each column other than "Sequence" is. 
art_metadata <- data.frame(
  SampleID = colnames(art_f)[-1],
  Condition = rep(c("dmso", "smg1i"), each = 3),  # 3 reps
  Replicate = rep(1:3, times = 2)  # 3 reps 
)

arti_metadata <- data.frame(
  SampleID = colnames(arti_f)[-1],
  Condition = rep(c("dmso", "smg1i"), each = 3),  # 3 reps
  Replicate = rep(1:3, times = 2)  # 3 reps 
)

#Create a deseq object and perform analysis for art and arti tables 
##ART DESeq analysis 
#re-level art
art_metadata$Condition <- relevel(factor(art_metadata$Condition), ref = "smg1i")

#generate a list of all the sequences in art
tgg_art <- subset(art_f,
                    substr(Sequence, 7, 9) %in% "TGG")
tgg_list_art<- tgg_art$Sequence

#design experiment for DESeq
art_dds <- DESeqDataSetFromMatrix(countData=art_f, 
                              colData=art_metadata, 
                              design=~Condition, tidy = TRUE)
#Set TGG
art_dds<- estimateSizeFactors(art_dds, 
                               controlGenes = rownames(art_dds) %in%
                                 tgg_list_art)
art_dds <- DESeq(art_dds)
art_results<- results(art_dds)

#histogram to make sure things make sense
hist(art_results$log2FoldChange, breaks = 80)

#overlaying histograms with specific sequences
art_results_d<- as.data.frame(art_results) #converting to dataframe
art_results_d$Sequence<- rownames(art_results_d) #making rownames part of the dataframe
rownames(art_results_d) <- NULL #making rownames null


##ARTi DESeq analysis
#re-level arti
arti_metadata$Condition <- relevel(factor(arti_metadata$Condition), ref = "smg1i")

#generate a list of all the sequences in arti 
tgg_arti <- subset(arti_f,
                    substr(Sequence, 7, 9) %in% "TGG")
tgg_list_arti<- tgg_arti$Sequence

#design experiment for DESeq
arti_dds <- DESeqDataSetFromMatrix(countData=arti_f, 
                              colData=arti_metadata, 
                              design=~Condition, tidy = TRUE)
#Set TGG
arti_dds<- estimateSizeFactors(arti_dds, 
                               controlGenes = rownames(arti_dds) %in%
                                 tgg_list_arti)
arti_dds <- DESeq(arti_dds)
arti_results<- results(arti_dds)

#histogram to make sure things make sense
hist(arti_results$log2FoldChange, breaks = 80)

#overlaying histograms with specific sequences
arti_results_d<- as.data.frame(arti_results) #converting to dataframe
arti_results_d$Sequence<- rownames(arti_results_d) #making rownames part of the dataframe
rownames(arti_results_d) <- NULL #making rownames null


write_csv(art_results_d, "DESeq_DNA_ART.csv")
write_csv(arti_results_d, "DESeq_DNA_ARTi.csv")

```

```{r linear modeling}
d <- DESeq_RNA_ART %>% filter(StopCodon != "TGG")
di<- DESeq_RNA_ARTi %>% filter(StopCodon != "TGG")
colnames(d)

library(broom)

DESeq_RNA_ART$StopCodon <- relevel(x = factor(DESeq_RNA_ART$StopCodon), ref = "TGG")
fit_stop<- lm(log2FoldChange ~ StopCodon, data = DESeq_RNA_ART)
tf_stop <- tidy(fit_stop)[-1,]
tf_stop$sig <- ifelse(tf_stop$p.value < 0.01, "sig", "n.s.")

fit_stop_i<- lm(log2FoldChange ~ StopCodon, data = DESeq_RNA_ARTi)
tf_stop_i <- tidy(fit_stop_i)[-1,]
tf_stop_i$sig <- ifelse(tf_stop_i$p.value < 0.01, "sig", "n.s.")

d$AA2 <- relevel(x = factor(d$AA2), ref = "Ala")
fit_AA2 <- lm(log2FoldChange ~ AA2, data = d)
tf_AA2 <- tidy(fit_AA2)[-1,]
tf_AA2$sig <- ifelse(tf_AA2$p.value < 0.01, "sig", "n.s.")
tf_AA2$AA<- substr(tf_AA2$term,4,6)
ggdotchart(data = tf_AA2, x = "AA", y = "estimate", color = "sig", size = 5) + 
  xlab("-1 Amino Acid")+
  ylab("Coefficient")+
  ylim(-0.5,0.4)+
  scale_color_manual( values = c("sig" = "#CF2F2E", "n.s." = "black"), name = "sig")+
  theme(text = element_text(size=20),
        axis.text.x = element_text(angle = 90, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" )



arti_results_d$StopCodon <- relevel(x = factor(arti_results_d$StopCodon), ref = "TGG")
fit_stop<- lm(log2FoldChange ~ StopCodon, data = art_results_d)
tf_stop <- tidy(fit_stop)[-1,]
tf_stop$sig <- ifelse(tf_stop$p.value < 0.01, "sig", "n.s.")
tf_stop$Stp<- substr(tf_stop$term, 10,12)
ggdotchart(data = tf_stop, x = "Stp", y = "estimate", color = "sig", size=7)+
  xlab("Stop Codon")+
  ylab("Coefficient")+
  ylim(-2.5,0)+
  scale_color_manual( values = c("sig" = "#CF2F2E", "n.s." = "black"), name = "sig")+
  theme(text = element_text(size=24),
        axis.text.x = element_text(angle = 0, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" )

d$AA1 <- relevel(x = factor(d$AA1), ref = "Ala")
fit_AA1 <- lm(log2FoldChange ~ AA1, data = di)
tf_AA1 <- tidy(fit_AA1)[-1,]
tf_AA1$sig <- ifelse(tf_AA1$p.value < 0.01, "sig", "n.s.")
tf_AA1$AA<- substr(tf_AA1$term,4,6)
ggdotchart(data = tf_AA1, x = "AA", y = "estimate", color = "sig", size = 5) + 
  xlab("-2 Amino Acid")+
  ylab("Coefficient")+
  ylim(-0.5,0.4)+
  scale_color_manual( values = c("sig" = "#CF2F2E", "n.s." = "black"), name = "sig")+
  theme(text = element_text(size=20),
        axis.text.x = element_text(angle = 90, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" )

fit_lnt <- lm(log2FoldChange ~ LastNt, data = d)
tf_lnt <- tidy(fit_lnt)[-1,]
tf_lnt$sig <- ifelse(tf_lnt$p.value < 0.01, "sig", "n.s.")
tf_lnt$lnt<- substr(tf_lnt$term,7,7)
ggdotchart(data = tf_lnt, x = "lnt", y = "estimate", color = "sig", size = 7) + 
  xlab("LastNt")+
  ylab("Coefficient")+
  ylim(-0.5,0.4)+
  scale_color_manual( values = c("sig" = "#CF2F2E", "n.s." = "black"), name = "sig")+
  theme(text = element_text(size=24),
        axis.text.x = element_text(angle = 0, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" )



fit_AA1_2 <- lm(log2FoldChange ~ AA1+AA2, data = d)
tf_AA1_2 <- tidy(fit_AA1_2)[-1,]
tf_AA1_2$sig <- tf_AA1_2$p.value < 0.01
tf_AA1_2$AAno<- substr(tf_AA1_2$term,1,3)
tf_AA1_2$AA<- substr(tf_AA1_2$term,4,6)

ggdotchart(data = tf_AA1_2, x = "AA", y = "estimate", color = "sig", size = 3) + 
  geom_hline(yintercept = 0, linetype="dashed" )+
  facet_wrap(~AAno)
library(ggrepel)
library(GGally)

#-1 Codon fit ART
fit_c2<- lm(log2FoldChange~ Codon2, data=d)
tidy_c2<- tidy(fit_c2)[-1,]
tidy_c2$C2 <- substring(tidy_c2$term,7,9)
tidy_c2$AA2 <- codon_to_aa(tidy_c2$C2, seq_type = "DNA")
tidy_c2$sig <- ifelse(tidy_c2$p.value < 0.01, "*", "")

heatmap_plot_c2 <- ggplot(tidy_c2, aes(x = C2, y=1, fill = estimate)) +
  geom_tile() +
  labs(x = "Codon2", fill = "coefficient") +
  theme_minimal()+
  facet_wrap(~AA2, scales="free_x", nrow=5)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),  
        panel.grid = element_blank(),  
        )+
  geom_text(aes(label = format(sig, nsmall = 3)))+
  scale_fill_gradient(low = "#CF2F2E", high = "#0726B9")+
  ggtitle("-1 Codon: EJC-Indep")

#-2 codon fit ART
fit_c1<- lm(log2FoldChange~ Codon1, data=d)
tidy_c1<- tidy(fit_c1)[-1,]
tidy_c1$C1 <- substring(tidy_c1$term,7,9)
tidy_c1$AA1 <- codon_to_aa(tidy_c1$C1, seq_type = "DNA")
tidy_c1$sig <- ifelse(tidy_c1$p.value < 0.01, "*", "")

heatmap_plot_c1 <- ggplot(tidy_c1, aes(x = C1, y=1, fill = estimate)) +
  geom_tile() +
  labs(x = "Codon1", fill = "coefficient") +
  theme_minimal()+
  facet_wrap(~AA1, scales="free_x", nrow=5)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),  
        panel.grid = element_blank(),  
        )+
  geom_text(aes(label = format(sig, nsmall = 3)))+
  scale_fill_gradient(low = "#CF2F2E", high = "#0726B9")+
  ggtitle("-2 Codon: EJC-Indep")

#-1 Codon fit ARTi
fit_c2_i<- lm(log2FoldChange~ Codon2, data=di)
tidy_c2_i<- tidy(fit_c2_i)[-1,]
tidy_c2_i$C2 <- substring(tidy_c2_i$term,7,9)
tidy_c2_i$AA2 <- codon_to_aa(tidy_c2_i$C2, seq_type = "DNA")
tidy_c2_i$sig <- ifelse(tidy_c2_i$p.value < 0.01, "*", "")

heatmap_plot_c2_i <- ggplot(tidy_c2_i, aes(x = C2, y=1, fill = estimate)) +
  geom_tile() +
  labs(x = "Codon2", fill = "coefficient") +
  theme_minimal()+
  facet_wrap(~AA2, scales="free_x", nrow=5)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),  
        panel.grid = element_blank(),  
        )+
  geom_text(aes(label = format(sig, nsmall = 3)))+
  scale_fill_gradient(low = "#CF2F2E", high = "#0726B9")+
  ggtitle("-1 Codon: EJC-Enh")

#-2 codon fit ARTi
fit_c1_i<- lm(log2FoldChange~ Codon1, data=di)
tidy_c1_i<- tidy(fit_c1_i)[-1,]
tidy_c1_i$C1 <- substring(tidy_c1_i$term,7,9)
tidy_c1_i$AA1 <- codon_to_aa(tidy_c1_i$C1, seq_type = "DNA")
tidy_c1_i$sig <- ifelse(tidy_c1_i$p.value < 0.01, "*", "")

heatmap_plot_c1_i <- ggplot(tidy_c1_i, aes(x = C1, y=1, fill = estimate)) +
  geom_tile() +
  labs(x = "Codon1", fill = "coefficient") +
  theme_minimal()+
  facet_wrap(~AA1, scales="free_x", nrow=5)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1),  
        panel.grid = element_blank(),  
        )+
  geom_text(aes(label = format(sig, nsmall = 3)))+
  scale_fill_gradient(low = "#CF2F2E", high = "#0726B9")+
  ggtitle("-2 Codon: EJC-Enh")


fg<- filter(tf_AA2_c2, term =="AA2Gly"|
             term =="Codon2GGG"|
             term =="Codon2GGC"|
             term =="Codon2GGA"
             #term =="Codon2GGT"
            )

fy<- filter(tf_AA2_c2, term =="AA2Phe"|
             term =="Codon2TTC"
              #term =="Codon2GGC"|
              #term =="Codon2GGT"|
              #term =="Codon2GGA"
            )

ggdotchart(data = fy, x = "term", y = "estimate", color = "sig", size = 7) + 
  xlab("Amino Acid or Codon")+
  ylab("Coefficient")+
  ylim(-0.5,0.4)+
  scale_color_manual( values = c("TRUE" = "#CF2F2E", "FALSE" = "black"), name = "sig")+
  theme(text = element_text(size=24),
        axis.text.x = element_text(angle = 0, hjust = 0.5, face="bold"),
        legend.position = "none")+
  geom_hline(yintercept = 0, linetype="dashed" )+
  scale_x_discrete(limits=c("AA2Phe", "Codon2TTC"))

lrtest(fit_AA2_c2,fit_AA2) 
lrtest(fit_AA1_2,fit_AA2)
lrtest(fit_AA2_c2,fit_AA1) 

glance(fit_AA2_c2)



d[sample(nrow(d),1000),]

set.seed(33)
testfit <- lm(log2FoldChange ~ AA2+Codon2+Codon1+AA1, data = d[sample(nrow(d),1000),])

glance(testfit)


d %>% filter(AA2=="Gly") %>%
  nest(data = -AA2) %>%
  mutate(
    fit = map(data, ~ lm(log2FoldChange ~ Codon2, data = .x)),
    tidied = map(fit, tidy)
  ) %>%
  unnest(tidied)
```



```{r preparing data for random forest}
##Adding properties to NMD sequences
#CU count ART
ct <- c("C", "T")
gc <-  c("G", "C")

d$ct_percent <- sapply(d$Sequence, function(text) {
  ctcount <- (sum(
    sapply(ct, 
           function(char) {nchar(gsub(paste0("[^", char, "]"), "", text))}
           )
) /10)*100
  })

d$gc_percent <- sapply(d$Sequence, function(text) {
  gccount <- (sum(
    sapply(gc, 
           function(char) {nchar(gsub(paste0("[^", char, "]"), "", text))}
           )
)/10)*100
  })

#ARTi
di$ct_percent <- sapply(di$Sequence, function(text) {
  ctcount <- (sum(
    sapply(ct, 
           function(char) {nchar(gsub(paste0("[^", char, "]"), "", text))}
           )
) /10)*100
  })

di$gc_percent <- sapply(di$Sequence, function(text) {
  gccount <- (sum(
    sapply(gc, 
           function(char) {nchar(gsub(paste0("[^", char, "]"), "", text))}
           )
)/10)*100
  })

#ART prepping other factors
#need to import codon properties, csc_bazinni, peptide hydrolysis, and phyprop
codonproperties<- merge(codonproperties, csc_bazzini, by="Codon")
PepHydRate <- PepHydRate[,c(1:5)]

d_all<- left_join(d, aa_phyprop, by = c("AA1"="Amino Acid"))
labels <- c("Hydropathy", "Charge", "pKaNH2", "pKaCOOH", "pKR", "Solubility") # Labels for the corresponding columns
colnames(d_all)[16:21] <- paste0(labels, "_AA1")

d_all<- left_join(d_all, aa_phyprop, by = c("AA2"="Amino Acid"))
colnames(d_all)[22:27] <- paste0(labels, "_AA2")

d_all<- left_join(d_all, codonproperties, by = c("Codon1"="Codon"))
labels_codon<- c("AA", "Frequency.1000", "tRNA binding", "CSC_293T")
colnames(d_all)[28:31] <- paste0(labels_codon, "_C1")

d_all<- left_join(d_all, codonproperties, by = c("Codon2"="Codon"))
colnames(d_all)[32:35] <- paste0(labels_codon, "_C2")

d_all<- left_join(d_all, PepHydRate, by=("AA2"))


d_all<- d_all %>% select(-c("AA_C1", "AA_C2", 
                            "baseMean", "lfcSE", "stat", "pvalue", "padj",
                            ))

colnames(d_all)[colnames(d_all) == "log2FoldChange"] <- "log2FC.RNA"

d_all <- d_all %>% arrange(log2FC.RNA)

#RF1 and RF2 density plots
#RF1
d_pep<- left_join(d, PepHydRate, by=("AA2"))
colnames(d_pep)[colnames(d_pep) == "log(RF1 rate)"] <- "lgrf1"
d_pep_fast<- subset(d_pep, lgrf1>1)
d_pep_slow<- subset(d_pep, lgrf1<= -0.5)

d_pep_rf1 <- ggplot()+
  geom_histogram(data=d_pep, aes(x = log2FoldChange, y=..count..), fill ="grey", alpha = 1, binwidth = 0.1)+
  geom_density(data=d_pep_fast, aes(x = log2FoldChange, y=..count..), color = "#0020B7", adjust = 2)+
  geom_density(data=d_pep_slow, aes(x = log2FoldChange, y=..count..), color = "#CD3E34", adjust = 2)+
  theme_classic()+
  ggtitle("PepHydRate_RF1")

d_pep_fast$pepcat<- "Fast"
d_pep_slow$pepcat<- "Slow"
y<- rbind(d_pep_fast, d_pep_slow)
wilcox.test(y$log2FoldChange ~ y$pepcat)

#RF2
colnames(d_pep)[colnames(d_pep) == "log(RF2 rate)"] <- "lgrf2"
d_pep_fast_r2<- subset(d_pep, lgrf2>1.25)
d_pep_slow_r2<- subset(d_pep, lgrf2<= 0)

d_pep_rf2 <- ggplot()+
  geom_histogram(data=d_pep, aes(x = log2FoldChange), fill ="grey", alpha = 1, binwidth = 0.1)+
  geom_density(data=d_pep_fast_r2, aes(x = log2FoldChange, y=..count..), color = "#0020B7", adjust = 2)+
  geom_density(data=d_pep_slow_r2, aes(x = log2FoldChange, y=..count..), color = "#CD3E34", adjust = 2)+
  theme_classic()+
  ggtitle("PepHydRate_RF2")

d_pep_fast_r2$pepcat<- "Fast"
d_pep_slow_r2$pepcat<- "Slow"
z<- rbind(d_pep_fast_r2, d_pep_slow_r2)
wilcox.test(z$log2FoldChange ~ z$pepcat)

#ARTi prepping other factors
di_all<- left_join(di, aa_phyprop, by = c("AA1"="Amino Acid"))
colnames(di_all)[16:21] <- paste0(labels, "_AA1")

di_all<- left_join(di_all, aa_phyprop, by = c("AA2"="Amino Acid"))
colnames(di_all)[22:27] <- paste0(labels, "_AA2")

di_all<- left_join(di_all, codonproperties, by = c("Codon1"="Codon"))
colnames(di_all)[28:31] <- paste0(labels_codon, "_C1")

di_all<- left_join(di_all, codonproperties, by = c("Codon2"="Codon"))
colnames(di_all)[32:35] <- paste0(labels_codon, "_C2")

di_all<- left_join(di_all, PepHydRate, by=("AA2"))

di_all<- di_all %>% select(-c("AA_C1", "AA_C2", 
                            "baseMean", "lfcSE", "stat", "pvalue", "padj",
                            ))

colnames(di_all)[colnames(di_all) == "log2FoldChange"] <- "log2FC.RNA"

di_all <- di_all %>% arrange(log2FC.RNA)


## Selecting top and bottom 3000 sequences 
#ART
d_highnmd<- d_all[1:3000,]
d_highnmd$NMD<- "strong nmd"

d_nmdesc <- tail(d_all, 3000)
d_nmdesc$NMD <- "nmd escape"

d_art_forest<- rbind(d_highnmd, d_nmdesc)

write_csv(d_art_forest, "art_rf.csv")

#ARTi
di_highnmd<- di_all[1:3000,]
di_highnmd$NMD<- "strong nmd"

di_nmdesc <- tail(di_all, 3000)
di_nmdesc$NMD <- "nmd escape"

di_arti_forest<- rbind(di_highnmd, di_nmdesc)

write_csv(di_arti_forest, "arti_rf.csv")






d_hnmd<- d$Sequence[1:3000]
d_ctrl<- d$Sequence[13981:16980]
d_esc<- tail(d$Sequence,3000)

write_fasta_no_headers <- function(sequences, output_file) {
  # Open the file for writing
  file_conn <- file(output_file, "w")
  
  # Initialize an index for the sequences
  sequence_index <- 1
  
  # Write each sequence to the file with a unique identifier
  for (sequence in sequences) {
    identifier <- paste0("Seq_", sequence_index)
    writeLines(paste0(">", identifier), file_conn)
    writeLines(sequence, file_conn)
    sequence_index <- sequence_index + 1
  }
  
  # Close the file
  close(file_conn)
  
  cat("FASTA file created:", output_file, "\n")
}


write_fasta_no_headers(d_hnmd, "highnmd_art.fasta")
write_fasta_no_headers(d_ctrl, "ctrl_art.fasta")
write_fasta_no_headers(d_esc, "nmdesc_art.fasta") 

di<- di%>%arrange(log2FoldChange)
di_hnmd<- di$Sequence[1:3000]
di_ctrl<- di$Sequence[17128:20127]
di_esc<- tail(di$Sequence,3000)

write_fasta_no_headers(di_hnmd, "highnmd_arti.fasta")
write_fasta_no_headers(di_ctrl, "ctrl_arti.fasta")
write_fasta_no_headers(di_esc, "nmdesc_arti.fasta") 

```

